service/history/mutableStateBuilder.go

```diff
@@ -3276,3 +3404,27 @@ func (e *mutableStateBuilder) CreateActivityRetryTimer(ai *persistence.ActivityI
 func (e *mutableStateBuilder) GetContinueAsNew() *persistence.CreateWorkflowExecutionRequest {
 	return e.continueAsNew
 }
+
+func (e *mutableStateBuilder) checkMutability(actionTag tag.Tag) error {
+	if !e.IsWorkflowExecutionRunning() {
+		e.logger.Warn(
+			mutableStateInvalidHistoryActionMsg,
+			tag.WorkflowEventID(e.GetNextEventID()),
+			tag.ErrorTypeInvalidHistoryAction,
+			tag.WorkflowState(e.executionInfo.State),
+			actionTag,
+		)
+		return ErrWorkflowFinished
+	}
+	return nil
+}
+
+func (e *mutableStateBuilder) createServerError(actionTag tag.Tag) error {
+	return &workflow.InternalServiceError{Message: actionTag.Field().String}
```

Comment:
This message will look weird. At least say XXX operation failed.

Reply:
Also createServerError -> createInternalServerError

Reply:
sure